omnipaxos/src/storage.rs

```diff
@@ -97,34 +97,149 @@ where
+    AppendEntries(Vec<T>),
+    ///
+    AppendOnPrefix(usize, Vec<T>),
+}
+
+/// transaction object for InternalStorage
+pub(crate) struct InternalStorageTxn<T: Entry> {
+    state_ops: Vec<StateOp>, // StateOp is an enum for modifying accepted_round, decided_idx, etc.
+    append_op: Vec<AppendOp<T>>,       // AppendOp is an enum for append, append_on_prefix, etc.
+    trim_op: Option<usize>,
+    stopsign: Option<Option<StopSign>>,
+    snapshot: Option<(usize, T::Snapshot)>  // snapshot index and snapshot
+}
+
+impl<T: Entry> InternalStorageTxn<T> {
+    /// TODO: docs
+    pub(crate) fn new() -> Self {
+        Self {
+            state_ops: vec![],
+            append_op: vec![],
+            trim_op: None,
+            stopsign: None,
+            snapshot: None,
+        }
+    } 
+
+    ///
+    pub(crate) fn set_promise(&mut self, promise: Ballot) {
+        self.state_ops.push(StateOp::SetPromise(promise))
+    }
+    ///
+    pub(crate) fn set_decided_idx(&mut self, idx: usize) {
+        self.state_ops.push(StateOp::SetDecidedIndex(idx))
+    }
+    ///
+    pub(crate) fn set_accepted_round(&mut self, round: Ballot) {
+        self.state_ops.push(StateOp::SetAcceptedRound(round))
+    }
+    ///
+    pub(crate) fn set_compacted_idx(&mut self, idx: usize) {
+        self.state_ops.push(StateOp::SetCompactedIdx(idx))
+    }
+    ///
+    pub(crate) fn append_entry(&mut self, entry: T) {
+        self.append_op.push(AppendOp::AppendEntry(entry))
+    }
+    ///
+    pub(crate) fn append_entries(&mut self, entries: Vec<T>) {
+        self.append_op.push(AppendOp::AppendEntries(entries))
+    }
+    ///
+    pub(crate) fn append_on_prefix(&mut self, from_idx: usize, entries: Vec<T>) {
+        self.append_op.push(AppendOp::AppendOnPrefix(from_idx, entries))
+    }
+    ///
+    pub(crate) fn trim(&mut self, from_idx: usize) {
+        self.trim_op = Some(from_idx)
+    }
+    ///
+    pub(crate) fn set_stopsign(&mut self, ss: Option<StopSign>) {
+        self.stopsign = Some(ss)
+    }
+    ///
+    pub(crate) fn set_snapshot(&mut self, compacted_idx: usize, snap: T::Snapshot) {
+        self.snapshot = Some((compacted_idx, snap))
+    }
+}
+
+
+/// TODO: docs
+#[derive(Debug)]
+pub enum StorageOp<T: Entry> {
+    ///
+    AppendEntry(T),
+    ///
+    AppendEntries(Vec<T>),
+    ///
+    AppendOnPrefix(usize, Vec<T>),
+    ///
+    SetPromise(Ballot),
+    ///
+    SetDecidedIndex(usize),
+    ///
+    SetAcceptedRound(Ballot),
+    ///
+    SetCompactedIdx(usize),
+    ///
+    Trim(usize),
+    ///
+    SetStopsign(Option<StopSign>),
+    ///
+    SetSnapshot(Option<T::Snapshot>),
+}
+
 /// Trait for implementing the storage backend of Sequence Paxos.
 pub trait Storage<T>
 where
     T: Entry,
 {
+    /// Atomically perform and persist all write operations in order.
+    /// Returns an Error if the transaction was not committed (so it was rolled back).
+    fn write_batch(&mut self, batch: Vec<StorageOp<T>>) -> StorageResult<()>;
```

Comment:
rename to `do_txn` or `perform_ops_atomically` to not confuse with the batched entries.

Reply:
please fix

Reply:
All the databases I looked at had different terminology for write batches and transactions. Write batches being what we want here (group of write operations to execute atomically) and transactions being what we specifically designed around not needing (groups of read and write operations where reads may need to some the effects of some writes).

Reply:
Yes. We're not a database so there is no need to get stuck up on terminology, especially if we're already using "batch" to refer to something else. 

I also disagree that "write batch" is a term that is generally perceived as being atomic (in contrast to a transaction). Therefore the name should include "transaction" or "atomic" in it.